internal/controller/promotions/promotions.go

```diff
@@ -158,205 +168,77 @@ func (r *reconciler) Reconcile(
+	if promo.Status.Phase == v1alpha1.PromotionPhaseRunning {
+		// anything we've already marked Running, we allow it to continue to reconcile
+	} else if promo.Status.Phase.IsTerminal() {
+		// if promo is already finished, nothing to do
+		return result, nil
+	} else {
+		// promo is Pending. Try to activate it.
+		if !r.pqs.tryActivate(ctx, promo) {
+			// It wasn't our turn. Mark this promo as pending (if it wasn't already)
+			if promo.Status.Phase != v1alpha1.PromotionPhasePending {
+				err = kubeclient.PatchStatus(ctx, r.kargoClient, promo, func(status *kargoapi.PromotionStatus) {
+					status.Phase = v1alpha1.PromotionPhasePending
+				})
+				return result, err
+			}
+			return result, nil
+		}
+	}
 
-	updateErr := kubeclient.PatchStatus(ctx, r.kargoClient, promo, func(status *kargoapi.PromotionStatus) {
-		*status = newStatus
+	logger = logger.WithFields(log.Fields{
+		"stage":   promo.Spec.Stage,
+		"freight": promo.Spec.Freight,
 	})
-	if updateErr != nil {
-		logger.Errorf("error updating Promotion status: %s", updateErr)
+	logger.Debug("executing Promotion")
+
+	// Update promo status as Running to give visibility in UI
+	if promo.Status.Phase != v1alpha1.PromotionPhaseRunning {
+		if err = kubeclient.PatchStatus(ctx, r.kargoClient, promo, func(status *kargoapi.PromotionStatus) {
+			status.Phase = v1alpha1.PromotionPhaseRunning
+		}); err != nil {
+			return result, err
+		}
 	}
 
-	// If we had no error, but couldn't update, then we DO have an error. But we
-	// do it this way so that a failure to update is never counted as THE failure
-	// when something else more serious occurred first.
-	if err == nil {
-		err = updateErr
-	}
+	promoCtx := logging.ContextWithLogger(ctx, logger)
 
-	// Controller runtime automatically gives us a progressive backoff if err is
-	// not nil
-	return result, err
-}
+	phase := kargoapi.PromotionPhaseSucceeded
+	phaseError := ""
 
-// initializeQueues lists all Promotions and adds them to relevant priority
-// queues. This is intended to be invoked ONCE and the caller MUST ensure that.
-// It is also assumed that the caller has already obtained a lock on
-// promoQueuesByStageMu.
-func (r *reconciler) initializeQueues(ctx context.Context) error {
-	promos := kargoapi.PromotionList{}
-	if err := r.kargoClient.List(ctx, &promos); err != nil {
-		return errors.Wrap(err, "error listing promotions")
-	}
-	logger := logging.LoggerFromContext(ctx)
-	for _, p := range promos.Items {
-		promo := p // This is to sidestep implicit memory aliasing in this for loop
-		if promo.Status.Phase.IsTerminal() {
-			continue
-		}
-		if promo.Status.Phase == "" {
-			if err := kubeclient.PatchStatus(ctx, r.kargoClient, &promo, func(status *kargoapi.PromotionStatus) {
-				status.Phase = kargoapi.PromotionPhasePending
-			}); err != nil {
-				return errors.Wrapf(
-					err,
-					"error updating status of Promotion %q in namespace %q",
-					promo.Name,
-					promo.Namespace,
-				)
+	func() {
```

Comment:
What's the reason for the anonymous function declaration here? As far as I can see, what's inside this block functions the same with or without it.

Reply:
Nevermind. I get it now.

Reply:
A comment explaining it might help, but otherwise 👍